daemon/graphdriver/overlay2/check.go

```diff
@@ -22,11 +23,23 @@ import (
 // directory or the kernel enable CONFIG_OVERLAY_FS_REDIRECT_DIR.
 // When these exist naive diff should be used.
 //
-// When running in a user namespace, returns errRunningInUserNS
-// immediately.
+// When running in a user namespace before kernel 5.11, returns
+// errRunningInUserNS immediately. In kernel 5.11 and later, we
+// check support as usual with some user namespace differences.
 func doesSupportNativeDiff(d string) error {
+	userxattr := false
 	if userns.RunningInUserNS() {
-		return errors.New("running in a user namespace")
+		if !kernel.CheckKernelVersion(5, 11, 0) {
```

Comment:
We generally do not want to introduce any new kernel version checks; this code should instead rely solely on the functional test.

Reply:
I have just commited a hopefully better solution which makes use of the existing `userxattr` check, since there is already a kernel version check in `NeedsUserXAttr()`. Unfortunately I cannot confirm if the new solution can cause problems in cases where `userxattr` is supported, while Native Overlay Diff isn't 🤔 I'll try to cross-check with previous bug reports to verify when I have the time.